### PR TITLE
To fix memory leaks on CRUD

### PR DESCRIPTION
**Pull request by sadhyama**

### CODE REVIEW

File: src/crud_internal.c [hunk 1]
```diff
@@ -119,188 +124,174 @@ int createObject( wrp_msg_t *reqMsg , wrp_msg_t **response)
...
+		jsonData = NULL;
+	}
 
-			/* Get the 1st object */
-			obj[0] = strdup( child_ptr );
-			ParodusPrint( "parent is %s\n", obj[0] );
+	if(reqMsg->u.crud.dest !=NULL)
+	{
+		ParodusInfo("reqMsg->u.crud.dest is %s\n", reqMsg->u.crud.dest);
+		child_ptr = strtok(reqMsg->u.crud.dest , "/");
```
Comment (schmidtw) [inline src/crud_internal.c]: This could be NULL as there could be no trailing `/`.

File: src/crud_internal.c [hunk 2]
```diff
@@ -119,188 +124,174 @@ int createObject( wrp_msg_t *reqMsg , wrp_msg_t **response)
...
 
-			objlevel = i;
-			ParodusInfo( "Number of object level %d\n", objlevel );
+		objlevel = i;
+		ParodusInfo( "Number of object level %d\n", objlevel );
 
-			/* Valid request will be mac:14cfexxxx/parodus/tags/${name} which is objlevel 4 */
-			if(objlevel == 4)
+		/* Valid request will be mac:14cfexxxx/parodus/tags/${name} which is objlevel 4 */
+		if(objlevel == 4)
```
Comment (schmidtw) [inline src/crud_internal.c]: Should add verification check for `0 == strcmp(obj[1], "mac:112233445566")`, `0 == strcmp(obj[2], "parodus")`, `0 == strcmp(obj[3], "tags")` unless the check is done elsewhere.

File: src/crud_internal.c [hunk 3]
```diff
@@ -119,188 +124,174 @@ int createObject( wrp_msg_t *reqMsg , wrp_msg_t **response)
...
-					{
-						jsonPayloadSize = cJSON_GetArraySize( jsonPayload );
-						ParodusPrint( "jsonPayloadSize is %d\n", jsonPayloadSize );
-					}
-
-					if(jsonPayloadSize)
+					jsonPayloadSize = cJSON_GetArraySize( jsonPayload );
+					ParodusPrint( "jsonPayloadSize is %d\n", jsonPayloadSize );
+				}
+				if(jsonPayloadSize)
```
Comment (schmidtw) [inline src/crud_internal.c]: Should be in the `if(jsonPayload !=NULL)` block?

File: src/crud_internal.c [hunk 4]
```diff
@@ -119,188 +124,174 @@ int createObject( wrp_msg_t *reqMsg , wrp_msg_t **response)
...
+		ParodusInfo("reqMsg->u.crud.dest is %s\n", reqMsg->u.crud.dest);
+		child_ptr = strtok(reqMsg->u.crud.dest , "/");
 
-			free(destVal);
+		/* Get the 1st object */
+		obj[0] = strdup( child_ptr );
+		ParodusPrint( "parent is %s\n", obj[0] );
 
-			while( child_ptr != NULL)
+		while( child_ptr != NULL)
```
Comment (schmidtw) [inline src/crud_internal.c]: This can overflow the hard coded limit of 5 above.  Needs to be limited to same max.

File: src/crud_internal.c [hunk 5]
```diff
@@ -448,208 +438,202 @@ int retrieveObject( wrp_msg_t *reqMsg, wrp_msg_t **response )
 	char *str1 = NULL;
 	const char *parse_error = NULL;
 
-	cJSON *jsonresponse = cJSON_CreateObject();
+	if(reqMsg->u.crud.dest !=NULL)
+	{
+		ParodusInfo("reqMsg->u.crud.dest is %s\n", reqMsg->u.crud.dest);
```
Comment (schmidtw) [inline src/crud_internal.c]: The block that breaks the destination into the components should leverage: `wrp_get_msg_dest_element()`.

File: src/crud_internal.c (hunk 1 above)
Comment (sadhyama) [reply src/crud_internal.c]: dest parsing logic is modified to use wrp_get_msg_dest_element() and child_ptr check is currently removed.

File: src/crud_internal.c (hunk 2 above)
Comment (sadhyama) [reply src/crud_internal.c]: Currently this check is done internally once objlevel is valid.

File: src/crud_internal.c (hunk 3 above)
Comment (sadhyama) [reply src/crud_internal.c]: Incorporated this comment.

File: src/crud_internal.c (hunk 4 above)
Comment (sadhyama) [reply src/crud_internal.c]: dest parsing logic is modified to use wrp_get_msg_dest_element() and hence this overflow problem is solved.

File: src/crud_internal.c (hunk 5 above)
Comment (sadhyama) [reply src/crud_internal.c]: Incorporated. Used wrp_get_msg_dest_element() and also added single function parse_dest_elements_to_string() to get pre-separated strings for each crud operation.

wrp_get_msg_dest_element() method from wrp-c is modified to support next element as crud requires dest format as mac:44aaf59b184a/parodus/tags/test.
Submitted PR for this change.
https://github.com/Comcast/wrp-c/pull/62

**Comment by sadhyama** [comment]
Travis build test failures will pass once wrp-c PR is merged and use latest git tag.

https://github.com/Comcast/wrp-c/pull/62

**Pull request merged**